*** SQLiteLexer.g4.new	Mon Aug  7 21:18:34 2023
--- SQLiteLexer.g4	Sun Sep 17 20:27:45 2023
***************
*** 80,86 ****
  COMMIT_:            'COMMIT';
  CONFLICT_:          'CONFLICT';
  CONSTRAINT_:        'CONSTRAINT';
- COPY_:              'COPY';
  CREATE_:            'CREATE';
  CROSS_:             'CROSS';
  CURRENT_DATE_:      'CURRENT_DATE';
--- 80,85 ----
***************
*** 170,176 ****
  TO_:                'TO';
  TRANSACTION_:       'TRANSACTION';
  TRIGGER_:           'TRIGGER';
- TRUNCATE_:          'TRUNCATE';
  UNION_:             'UNION';
  UNIQUE_:            'UNIQUE';
  UPDATE_:            'UPDATE';
--- 169,174 ----
