*** SQLiteParser.g4.new	Mon Jun 26 23:59:31 2023
--- SQLiteParser.g4	Tue Jun 27 00:00:12 2023
***************
*** 47,53 ****
          | attach_stmt
          | begin_stmt
          | commit_stmt
-         | create_database_stmt
          | create_index_stmt
          | create_table_stmt
          | create_trigger_stmt
--- 47,52 ----
***************
*** 56,66 ****
          | delete_stmt
          | delete_stmt_limited
          | detach_stmt
!         | drop_database_stmt
!         | drop_index_stmt
!         | drop_table_stmt
!         | drop_trigger_stmt
!         | drop_view_stmt
          | insert_stmt
          | pragma_stmt
          | reindex_stmt
--- 55,61 ----
          | delete_stmt
          | delete_stmt_limited
          | detach_stmt
!         | drop_stmt
          | insert_stmt
          | pragma_stmt
          | reindex_stmt
***************
*** 115,137 ****
  ;
  
  create_index_stmt:
!     CREATE_ UNIQUE_? INDEX_ (if_not_exists)? (schema_name DOT)? index_name ON_ table_name OPEN_PAR
          indexed_column (COMMA indexed_column)* CLOSE_PAR (WHERE_ expr)?
  ;
  
  indexed_column: (column_name | expr) (COLLATE_ collation_name)? asc_desc?
  ;
  
- create_database_stmt:
-     CREATE_ DATABASE_ (if_not_exists)? database_name
- ;
- 
- if_not_exists:
-     IF_ NOT_ EXISTS_
- ;
- 
  create_table_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? TABLE_ (if_not_exists)? (
          schema_name DOT
      )? table_name (
          OPEN_PAR column_def (COMMA column_def)*? (COMMA table_constraint)* CLOSE_PAR (
--- 110,124 ----
  ;
  
  create_index_stmt:
!     CREATE_ UNIQUE_? INDEX_ (IF_ NOT_ EXISTS_)? (schema_name DOT)? index_name ON_ table_name OPEN_PAR
          indexed_column (COMMA indexed_column)* CLOSE_PAR (WHERE_ expr)?
  ;
  
  indexed_column: (column_name | expr) (COLLATE_ collation_name)? asc_desc?
  ;
  
  create_table_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? TABLE_ (IF_ NOT_ EXISTS_)? (
          schema_name DOT
      )? table_name (
          OPEN_PAR column_def (COMMA column_def)*? (COMMA table_constraint)* CLOSE_PAR (
***************
*** 203,209 ****
  ;
  
  create_trigger_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? TRIGGER_ (if_not_exists)? (
          schema_name DOT
      )? trigger_name (BEFORE_ | AFTER_ | INSTEAD_ OF_)? (
          DELETE_
--- 190,196 ----
  ;
  
  create_trigger_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? TRIGGER_ (IF_ NOT_ EXISTS_)? (
          schema_name DOT
      )? trigger_name (BEFORE_ | AFTER_ | INSTEAD_ OF_)? (
          DELETE_
***************
*** 215,227 ****
  ;
  
  create_view_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? VIEW_ (if_not_exists)? (
          schema_name DOT
      )? view_name (OPEN_PAR column_name (COMMA column_name)* CLOSE_PAR)? AS_ select_stmt
  ;
  
  create_virtual_table_stmt:
!     CREATE_ VIRTUAL_ TABLE_ (if_not_exists)? (schema_name DOT)? table_name USING_ module_name (
          OPEN_PAR module_argument (COMMA module_argument)* CLOSE_PAR
      )?
  ;
--- 202,214 ----
  ;
  
  create_view_stmt:
!     CREATE_ (TEMP_ | TEMPORARY_)? VIEW_ (IF_ NOT_ EXISTS_)? (
          schema_name DOT
      )? view_name (OPEN_PAR column_name (COMMA column_name)* CLOSE_PAR)? AS_ select_stmt
  ;
  
  create_virtual_table_stmt:
!     CREATE_ VIRTUAL_ TABLE_ (IF_ NOT_ EXISTS_)? (schema_name DOT)? table_name USING_ module_name (
          OPEN_PAR module_argument (COMMA module_argument)* CLOSE_PAR
      )?
  ;
***************
*** 260,303 ****
  
  drop_stmt:
      DROP_ object = (INDEX_ | TABLE_ | TRIGGER_ | VIEW_) (
!         if_exists
      )? (schema_name DOT)? any_name
  ;
  
- drop_database_stmt:
-     DROP_  DATABASE_ (
-         if_exists
-     )? database_name
- ;
- 
- drop_index_stmt:
-     DROP_  INDEX_ (
-         if_exists
-     )? (schema_name DOT)? any_name
- ;
- 
- drop_table_stmt:
-     DROP_  TABLE_ (
-         if_exists
-     )? (schema_name DOT)? any_name
- ;
- 
- drop_trigger_stmt:
-     DROP_  TRIGGER_ (
-         if_exists
-     )? (schema_name DOT)? any_name
- ;
- 
- drop_view_stmt:
-     DROP_  VIEW_ (
-         if_exists
-     )? (schema_name DOT)? any_name
- ;
- 
- if_exists:
-     IF_ NOT_ EXISTS_
- ;
- 
  /*
   SQLite understands the following binary operators, in order from highest to lowest precedence:
      ||
--- 247,256 ----
  
  drop_stmt:
      DROP_ object = (INDEX_ | TABLE_ | TRIGGER_ | VIEW_) (
!         IF_ EXISTS_
      )? (schema_name DOT)? any_name
  ;
  
  /*
   SQLite understands the following binary operators, in order from highest to lowest precedence:
      ||
***************
*** 863,872 ****
  // TODO: check all names below
  
  name:
-     any_name
- ;
- 
- database_name:
      any_name
  ;
  
--- 816,821 ----
